### PR TITLE
implicit-casts:false in flutter/lib/src/widgets

### PR DESCRIPTION
**Pull request by a14n**
## Description

Enable `implicit-casts:false` in `flutter/lib/src/widgets`

## Related Issues

None

## Tests

None because it's a refactoring.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/focus_manager.dart [hunk 1]
```diff
@@ -595,7 +595,7 @@ class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
   ///
   /// Use [nearestScope] to start at this node instead of above it.
   FocusScopeNode get enclosingScope {
-    return ancestors.firstWhere((FocusNode node) => node is FocusScopeNode, orElse: () => null);
+    return ancestors.whereType<FocusScopeNode>().firstWhere((_) => true, orElse: () => null);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/focus_manager.dart]: The `firstWhere((_) => true` part is strange.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -65,8 +65,8 @@ class ObjectKey extends LocalKey {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-    final ObjectKey typedOther = other;
-    return identical(value, typedOther.value);
+    return other is ObjectKey
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart]: Same as in the other PRs.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -314,8 +314,8 @@ class GlobalObjectKey<T extends State<StatefulWidget>> extends GlobalKey<T> {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-    final GlobalObjectKey<T> typedOther = other;
-    return identical(value, typedOther.value);
+    return other is GlobalObjectKey<T>
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/framework.dart]: Same as in the other PRs.

File: packages/flutter/lib/src/widgets/gesture_detector.dart [hunk 4]
```diff
@@ -1126,16 +1126,16 @@ class _DefaultSemanticsGestureDelegate extends SemanticsGestureDelegate {
   @override
   void assignSemantics(RenderSemanticsGestureHandler renderObject) {
     assert(!detectorState.widget.excludeFromSemantics);
-    final Map<Type, GestureRecognizer> recognizers = detectorState._recognizers;
+    T getRecognizer<T extends GestureRecognizer>() => detectorState._recognizers[T] as T;
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: This seems oddly complicated. Can just each _get*Handler do the cast?

File: packages/flutter/lib/src/widgets/icon_data.dart [hunk 5]
```diff
@@ -53,11 +53,11 @@ class IconData {
   bool operator ==(dynamic other) {
     if (runtimeType != other.runtimeType)
       return false;
-    final IconData typedOther = other;
-    return codePoint == typedOther.codePoint
-        && fontFamily == typedOther.fontFamily
-        && fontPackage == typedOther.fontPackage
-        && matchTextDirection == typedOther.matchTextDirection;
+    return other is IconData
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/icon_data.dart]: Same as elsewhere.

File: packages/flutter/lib/src/widgets/icon_theme_data.dart [hunk 6]
```diff
@@ -103,10 +103,10 @@ class IconThemeData extends Diagnosticable {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-    final IconThemeData typedOther = other;
-    return color == typedOther.color
-        && opacity == typedOther.opacity
-        && size == typedOther.size;
+    return other is IconThemeData
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/icon_theme_data.dart]: Same

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 7]
```diff
@@ -535,21 +535,21 @@ class MediaQueryData {
...
-        && typedOther.viewPadding == viewPadding
-        && typedOther.viewInsets == viewInsets
-        && typedOther.physicalDepth == physicalDepth
-        && typedOther.alwaysUse24HourFormat == alwaysUse24HourFormat
-        && typedOther.highContrast == highContrast
-        && typedOther.disableAnimations == disableAnimations
-        && typedOther.invertColors == invertColors
-        && typedOther.accessibleNavigation == accessibleNavigation
-        && typedOther.boldText == boldText;
+    return other is MediaQueryData
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: Same.

File: packages/flutter/lib/src/widgets/page_storage.dart [hunk 8]
```diff
@@ -48,12 +49,8 @@ class _StorageEntryIdentifier {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-    final _StorageEntryIdentifier typedOther = other;
-    for (int index = 0; index < keys.length; index += 1) {
-      if (keys[index] != typedOther.keys[index])
-        return false;
-    }
-    return true;
+    return other is _StorageEntryIdentifier
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/page_storage.dart]: Same.

File: packages/flutter/lib/src/widgets/page_storage.dart [hunk 9]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import '../../foundation.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/page_storage.dart]: foundation should be imported as package:foundation.dart and not with relative path.

File: packages/flutter/lib/src/widgets/shortcuts.dart [hunk 10]
```diff
@@ -89,8 +89,8 @@ class KeySet<T extends KeyboardKey> extends Diagnosticable {
     if (other.runtimeType != runtimeType) {
       return false;
     }
-    final KeySet<T> typedOther = other;
-    return setEquals<T>(_keys, typedOther._keys);
+    return other is KeySet<T>
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/shortcuts.dart]: Same.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 11]
```diff
@@ -2438,8 +2438,9 @@ class _TransformedRect {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-    final _TransformedRect typedOther = other;
-    return rect == typedOther.rect && transform == typedOther.transform;
+    return other is _TransformedRect
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Same.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 12]
```diff
@@ -2469,12 +2470,11 @@ class _InspectorOverlayRenderState {
   bool operator ==(dynamic other) {
     if (other.runtimeType != runtimeType)
       return false;
-
-    final _InspectorOverlayRenderState typedOther = other;
-    return overlayRect == typedOther.overlayRect
-        && selected == typedOther.selected
-        && listEquals<_TransformedRect>(candidates, typedOther.candidates)
-        && tooltip == typedOther.tooltip;
+    return other is _InspectorOverlayRenderState
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: Same.

File: packages/flutter/lib/src/widgets/widget_span.dart [hunk 13]
```diff
@@ -171,10 +171,10 @@ class WidgetSpan extends PlaceholderSpan {
       return false;
     if (super != other)
       return false;
-    final WidgetSpan typedOther = other;
-    return typedOther.child == child
-        && typedOther.alignment == alignment
-        && typedOther.baseline == baseline;
+    return other is WidgetSpan
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_span.dart]: Same

File: packages/flutter/lib/src/widgets/widget_span.dart [hunk 14]
```diff
@@ -171,10 +171,10 @@ class WidgetSpan extends PlaceholderSpan {
       return false;
     if (super != other)
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_span.dart]: (unrelated): This if looks buggy. I'll take a look at this...

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/widget_span.dart]: Never mind, I missread this.

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: Yes, but `.first` throws if the Iterable is empty. `firstWhere` is used only to have a default to null.

Here are the alternatives:

```dart

// using a explicit cast
FocusScopeNode get enclosingScope {
  return ancestors.firstWhere((FocusNode node) => node is FocusScopeNode, orElse: () => null) as FocusScopeNode;
}

// using firstWhere and orElse
FocusScopeNode get enclosingScope {
  return ancestors.whereType<FocusScopeNode>().firstWhere((_) => true, orElse: () => null);
}

// catching error
FocusScopeNode get enclosingScope {
  try {
    return ancestors.whereType<FocusScopeNode>().first;
  } on StateError {
    return null;
  }
}

// with spread
FocusScopeNode get enclosingScope {
  return <FocusScopeNode>[...ancestors.whereType<FocusScopeNode>(), null].first;
}

// with static extension methods
extension on Iterable<T> {
  T get firstOrNull  {
    try {
      return this.first;
    } on StateError {
      return null;
    }
  }
}
FocusScopeNode get enclosingScope {
  return ancestors.whereType<FocusScopeNode>().firstOrNull;
}
```

What would you prefer?

File: packages/flutter/lib/src/widgets/gesture_detector.dart (hunk 4 above)
Comment (a14n) [reply packages/flutter/lib/src/widgets/gesture_detector.dart]: ok, I will revert this part and go with a simple explicit cast.

File: packages/flutter/lib/src/widgets/page_storage.dart (hunk 9 above)
Comment (a14n) [reply packages/flutter/lib/src/widgets/page_storage.dart]: Done

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: They are all not great, I would prefer the "using a explicit cast" option, though.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/focus_manager.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/widgets/focus_manager.dart]: Changed with "using a explicit cast"

**Pull request merged**